<a href="https://colab.research.google.com/github/girdharriya/Data-Analytics-Project/blob/main/Datapreprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Assignment 1 - DATA PRE-PROCESSING**


In [ ]:
from google.colab import drive

In [ ]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#Installing the required libraries
import pandas as pd, numpy as np # For Data Manipulation
from sklearn.preprocessing import LabelEncoder, OrdinalEncoder # For Encoding Categorical Data [Nominal | Ordinal]
from sklearn.preprocessing import OneHotEncoder # For Creating Dummy Variables of Categorical Data [Nominal]
from sklearn.impute import SimpleImputer, KNNImputer # For Imputation of Missing Data
from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler # For Rescaling Data
from sklearn.model_selection import train_test_split # For Splitting Data into Training & Testing Sets

In [ ]:
#Loading the dataset
data=pd.read_csv('/content/healthcare-dataset-stroke-data.csv')

# **1.Treatment of Missing Data [Null, NaN, None]**

In [ ]:
#Analysing the shape of the dataset & displaying the first 5 rows
data.shape,data.head(5)

((5110, 12),
       id  gender   age  hypertension  heart_disease ever_married  \
 0   9046    Male  67.0             0              1          Yes   
 1  51676  Female  61.0             0              0          Yes   
 2  31112    Male  80.0             0              1          Yes   
 3  60182  Female  49.0             0              0          Yes   
 4   1665  Female  79.0             1              0          Yes   
 
        work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
 0        Private          Urban             228.69  36.6  formerly smoked   
 1  Self-employed          Rural             202.21   NaN     never smoked   
 2        Private          Rural             105.92  32.5     never smoked   
 3        Private          Urban             171.23  34.4           smokes   
 4  Self-employed          Rural             174.12  24.0     never smoked   
 
    stroke  
 0       1  
 1       1  
 2       1  
 3       1  
 4       1  )

In [ ]:
#Details of the dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [ ]:
#Converting String to Boolean Type
data['ever_married'] = data['ever_married'].replace({'Yes': 1, 'No': 0})

**1.1. Missing Data Information**

In [ ]:
#Checking for missing values columnwise
data_missing=data.isna().sum()
#Record-wise Missing Data Information (Top 5)
record_missing_data = data.isna().sum(axis=1).sort_values(ascending=False).head(5)
data_missing,record_missing_data

(id                     0
 gender                 0
 age                    0
 hypertension           0
 heart_disease          0
 ever_married           0
 work_type              0
 Residence_type         0
 avg_glucose_level      0
 bmi                  201
 smoking_status         0
 stroke                 0
 dtype: int64,
 680     1
 183     1
 1235    1
 170     1
 171     1
 dtype: int64)

In [ ]:
# Data Bifurcation
# Non-Categorical Data
df_noncat = data[['id','age','hypertension','heart_disease','ever_married','avg_glucose_level','bmi','stroke']]
# Categorical Data [Nominal | Ordinal]
df_cat = data[['gender', 'work_type','Residence_type', 'smoking_status']]

**1.2. Impute Missing Categorical Data [Nominal | Ordinal] using Descriptive Statistics : Central Tendency**

In [ ]:
# Strategy = median [When Odd Number of Categories Exists]
si_cat = SimpleImputer(missing_values=np.nan, strategy='most_frequent')
si_cat_fit = si_cat.fit_transform(df_cat)
# Missing Categorical Data Imputed Subset
df_cat_mdi = pd.DataFrame(si_cat_fit, columns=df_cat.columns)
df_cat_mdi , df_cat_mdi.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 4 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   gender          5110 non-null   object
 1   work_type       5110 non-null   object
 2   Residence_type  5110 non-null   object
 3   smoking_status  5110 non-null   object
dtypes: object(4)
memory usage: 159.8+ KB


(      gender      work_type Residence_type   smoking_status
 0       Male        Private          Urban  formerly smoked
 1     Female  Self-employed          Rural     never smoked
 2       Male        Private          Rural     never smoked
 3     Female        Private          Urban           smokes
 4     Female  Self-employed          Rural     never smoked
 ...      ...            ...            ...              ...
 5105  Female        Private          Urban     never smoked
 5106  Female  Self-employed          Urban     never smoked
 5107  Female  Self-employed          Rural     never smoked
 5108    Male        Private          Rural  formerly smoked
 5109  Female       Govt_job          Urban          Unknown
 
 [5110 rows x 4 columns],
 None)

**1.3. Impute Missing Non-Categorical Data using Descriptive Statistics : Central Tendency**

In [ ]:
# Strategy : mean | median | most_frequent | constant
si_noncat = SimpleImputer(missing_values=np.nan, strategy='mean')
si_noncat_fit = si_noncat.fit_transform(df_noncat)
# Missing Non-Categorical Data Imputed Subset using Simple Imputer
df_noncat_mdi_si = pd.DataFrame(si_noncat_fit, columns=df_noncat.columns)
df_noncat_mdi_si,  df_noncat_mdi_si.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   float64
 1   age                5110 non-null   float64
 2   hypertension       5110 non-null   float64
 3   heart_disease      5110 non-null   float64
 4   ever_married       5110 non-null   float64
 5   avg_glucose_level  5110 non-null   float64
 6   bmi                5110 non-null   float64
 7   stroke             5110 non-null   float64
dtypes: float64(8)
memory usage: 319.5 KB


(           id   age  hypertension  heart_disease  ever_married  \
 0      9046.0  67.0           0.0            1.0           1.0   
 1     51676.0  61.0           0.0            0.0           1.0   
 2     31112.0  80.0           0.0            1.0           1.0   
 3     60182.0  49.0           0.0            0.0           1.0   
 4      1665.0  79.0           1.0            0.0           1.0   
 ...       ...   ...           ...            ...           ...   
 5105  18234.0  80.0           1.0            0.0           1.0   
 5106  44873.0  81.0           0.0            0.0           1.0   
 5107  19723.0  35.0           0.0            0.0           1.0   
 5108  37544.0  51.0           0.0            0.0           1.0   
 5109  44679.0  44.0           0.0            0.0           1.0   
 
       avg_glucose_level        bmi  stroke  
 0                228.69  36.600000     1.0  
 1                202.21  28.893237     1.0  
 2                105.92  32.500000     1.0  
 3        

**1.4. Impute Missing Non-Categorical Data using K-Nearest Neighbors**

In [ ]:
# weights : uniform | distance | {User Defined}
ki = KNNImputer(n_neighbors=5, weights='uniform')
ki_fit = ki.fit_transform(df_noncat)
# Missing Non-Categorical Data Imputed Subset using KNN Imputer
df_noncat_mdi_ki = pd.DataFrame(ki_fit, columns=df_noncat.columns)
df_noncat_mdi_ki, df_noncat_mdi_ki.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   float64
 1   age                5110 non-null   float64
 2   hypertension       5110 non-null   float64
 3   heart_disease      5110 non-null   float64
 4   ever_married       5110 non-null   float64
 5   avg_glucose_level  5110 non-null   float64
 6   bmi                5110 non-null   float64
 7   stroke             5110 non-null   float64
dtypes: float64(8)
memory usage: 319.5 KB


(           id   age  hypertension  heart_disease  ever_married  \
 0      9046.0  67.0           0.0            1.0           1.0   
 1     51676.0  61.0           0.0            0.0           1.0   
 2     31112.0  80.0           0.0            1.0           1.0   
 3     60182.0  49.0           0.0            0.0           1.0   
 4      1665.0  79.0           1.0            0.0           1.0   
 ...       ...   ...           ...            ...           ...   
 5105  18234.0  80.0           1.0            0.0           1.0   
 5106  44873.0  81.0           0.0            0.0           1.0   
 5107  19723.0  35.0           0.0            0.0           1.0   
 5108  37544.0  51.0           0.0            0.0           1.0   
 5109  44679.0  44.0           0.0            0.0           1.0   
 
       avg_glucose_level    bmi  stroke  
 0                228.69  36.60     1.0  
 1                202.21  28.08     1.0  
 2                105.92  32.50     1.0  
 3                171.23  

**1.5. Missing Data Exclusion [MCAR | MAR (> 50%)]**

In [ ]:
# Dataset Used : df_cat_mdi | df_noncat_mdi_si | df_noncat_mdi_ki

# Excluding Empty Records (If Any)
df_cat_mdi.dropna(axis=0, how='all', inplace=True) # Categorical Data Subset
# Non-Categorical Data Subset
df_noncat_mdi_si.dropna(axis=0, how='all', inplace=True)
# Non-Categorical Data Subset
df_noncat_mdi_ki.dropna(axis=0, how='all', inplace=True)

In [ ]:
# Excluding Empty Variables (If Any)
df_cat_mdi.dropna(axis=1, how='all', inplace=True) # Categorical Data Subset
# Non-Categorical Data Subset
df_noncat_mdi_si.dropna(axis=1, how='all', inplace=True)
# Non-Categorical Data Subset
df_noncat_mdi_ki.dropna(axis=1, how='all', inplace=True)

In [ ]:
# Missing Categorical Treated Dataset
df_cat_mdt = df_cat_mdi.copy()
# Missing Non-Categorical Treated Dataset
df_noncat_mdt = df_noncat_mdi_si.copy()
# Missing Non-Categorical Treated Dataset
df_noncat_mdt = df_noncat_mdi_ki.copy()

# **2. Numeric Coding of Data**

***2.1. Numeric Encoding of Categorical Data [Nominal & Ordinal] ***

In [ ]:
df_cat_mdt_code = df_cat_mdt.copy()
# Using Pandas (Inferior)
df_cat_mdt_code_pd = df_cat_mdt_code.astype('category')
df_cat_mdt_code_pd['gender_code'] = df_cat_mdt_code_pd['gender'].cat.codes
df_cat_mdt_code_pd['work_type_code'] = df_cat_mdt_code_pd['work_type'].cat.codes
df_cat_mdt_code_pd['Residence_type_code'] = df_cat_mdt_code_pd['Residence_type'].cat.codes
df_cat_mdt_code_pd['smoking_status_code'] = df_cat_mdt_code_pd['smoking_status'].cat.codes
# (Missing Data Treated) Numeric Coded Categorical Dataset using Pandas
df_cat_mdt_code_pd

gender      work_type Residence_type   smoking_status  gender_code  \
0       Male        Private          Urban  formerly smoked            1   
1     Female  Self-employed          Rural     never smoked            0   
2       Male        Private          Rural     never smoked            1   
3     Female        Private          Urban           smokes            0   
4     Female  Self-employed          Rural     never smoked            0   
...      ...            ...            ...              ...          ...   
5105  Female        Private          Urban     never smoked            0   
5106  Female  Self-employed          Urban     never smoked            0   
5107  Female  Self-employed          Rural     never smoked            0   
5108    Male        Private          Rural  formerly smoked            1   
5109  Female       Govt_job          Urban          Unknown            0   

      work_type_code  Residence_type_code  smoking_status_code  
0                  2                    1                    1  
1                  3                    0                    2  
2                  2                    0                    2  
3                  2                    1                    3  
4                  3                    0                    2  
...              ...                  ...                  ...  
5105               2                    1                    2  
5106               3                    1                    2  
5107               3                    0                    2  
5108               2                    0                    1  
5109               0                    1                    0  

[5110 rows x 8 columns]

In [ ]:
# Using Scikit Learn : Label Encoder (Inferior)
le = LabelEncoder()
le_fit1 = le.fit_transform(df_cat_mdt_code['gender'])
le_fit2 = le.fit_transform(df_cat_mdt_code['work_type'])
le_fit3 = le.fit_transform(df_cat_mdt_code['Residence_type'])
le_fit4 = le.fit_transform(df_cat_mdt_code['smoking_status'])
df_gen_code = pd.DataFrame(le_fit1, columns=['gender_code']); df_gen_code
df_work_code = pd.DataFrame(le_fit2, columns=['work_type_code']); df_work_code
df_resi_code = pd.DataFrame(le_fit3, columns=['Residence_type_code']); df_resi_code
df_smok_code = pd.DataFrame(le_fit4, columns=['smoking_status_code']); df_smok_code
df_gen_work_code = df_gen_code.join(df_work_code); df_gen_work_code
df_gen_work_resi_code = df_gen_work_code.join(df_resi_code); df_gen_work_resi_code
df_gen_work_resi_smok_code = df_gen_work_resi_code.join(df_smok_code); df_gen_work_resi_smok_code
# (Missing Data Treated) Numeric Coded Categorical Dataset using Scikit Learn Label Encoder
df_cat_mdt_code_le = df_cat_mdt_code.join(df_gen_work_resi_smok_code); df_cat_mdt_code_le
df_cat_mdt_code_le = pd.merge(df_cat_mdt_code, df_gen_work_resi_smok_code, left_index=True, right_index=True); df_cat_mdt_code_le

gender      work_type Residence_type   smoking_status  gender_code  \
0       Male        Private          Urban  formerly smoked            1   
1     Female  Self-employed          Rural     never smoked            0   
2       Male        Private          Rural     never smoked            1   
3     Female        Private          Urban           smokes            0   
4     Female  Self-employed          Rural     never smoked            0   
...      ...            ...            ...              ...          ...   
5105  Female        Private          Urban     never smoked            0   
5106  Female  Self-employed          Urban     never smoked            0   
5107  Female  Self-employed          Rural     never smoked            0   
5108    Male        Private          Rural  formerly smoked            1   
5109  Female       Govt_job          Urban          Unknown            0   

      work_type_code  Residence_type_code  smoking_status_code  
0                  2                    1                    1  
1                  3                    0                    2  
2                  2                    0                    2  
3                  2                    1                    3  
4                  3                    0                    2  
...              ...                  ...                  ...  
5105               2                    1                    2  
5106               3                    1                    2  
5107               3                    0                    2  
5108               2                    0                    1  
5109               0                    1                    0  

[5110 rows x 8 columns]

In [ ]:
# Using Scikit Learn : Ordinal Encoder (Superior)
oe = OrdinalEncoder()
oe_fit = oe.fit_transform(df_cat_mdt_code)
df_cat_code_oe = pd.DataFrame(oe_fit, columns=['gender_code', 'work_type_code','Residence_type_code','smoking_status_code']); df_cat_code_oe
# (Missing Data Treated) Numeric Coded Categorical Dataset using Scikit Learn Ordinal Encoder
df_cat_mdt_code_oe = df_cat_mdt_code.join(df_cat_code_oe); df_cat_mdt_code_oe
df_cat_mdt_code_oe = pd.merge(df_cat_mdt_code, df_cat_code_oe, left_index=True, right_index=True); df_cat_mdt_code_oe


gender      work_type Residence_type   smoking_status  gender_code  \
0       Male        Private          Urban  formerly smoked          1.0   
1     Female  Self-employed          Rural     never smoked          0.0   
2       Male        Private          Rural     never smoked          1.0   
3     Female        Private          Urban           smokes          0.0   
4     Female  Self-employed          Rural     never smoked          0.0   
...      ...            ...            ...              ...          ...   
5105  Female        Private          Urban     never smoked          0.0   
5106  Female  Self-employed          Urban     never smoked          0.0   
5107  Female  Self-employed          Rural     never smoked          0.0   
5108    Male        Private          Rural  formerly smoked          1.0   
5109  Female       Govt_job          Urban          Unknown          0.0   

      work_type_code  Residence_type_code  smoking_status_code  
0                2.0                  1.0                  1.0  
1                3.0                  0.0                  2.0  
2                2.0                  0.0                  2.0  
3                2.0                  1.0                  3.0  
4                3.0                  0.0                  2.0  
...              ...                  ...                  ...  
5105             2.0                  1.0                  2.0  
5106             3.0                  1.0                  2.0  
5107             3.0                  0.0                  2.0  
5108             2.0                  0.0                  1.0  
5109             0.0                  1.0                  0.0  

[5110 rows x 8 columns]

***2.2. Creation of Dummy Variables from Categorical Data ***

In [ ]:
df_cat_mdt_code = df_cat_mdt.copy()
# Dummy Variable : gender
# Using Pandas (Superior)
df_mdt_gen_dummy_pd = pd.get_dummies(df_cat_mdt_code[['gender']]); df_mdt_gen_dummy_pd # (Missing Data Treated) Dummy Dataset using Pandas
df_mdt_cat_dummy_pd = pd.get_dummies(df_cat_mdt_code, columns = ['gender', 'work_type','Residence_type','smoking_status']); df_mdt_cat_dummy_pd


gender_Female  gender_Male  gender_Other  work_type_Govt_job  \
0                 0            1             0                   0   
1                 1            0             0                   0   
2                 0            1             0                   0   
3                 1            0             0                   0   
4                 1            0             0                   0   
...             ...          ...           ...                 ...   
5105              1            0             0                   0   
5106              1            0             0                   0   
5107              1            0             0                   0   
5108              0            1             0                   0   
5109              1            0             0                   1   

      work_type_Never_worked  work_type_Private  work_type_Self-employed  \
0                          0                  1                        0   
1                          0                  0                        1   
2                          0                  1                        0   
3                          0                  1                        0   
4                          0                  0                        1   
...                      ...                ...                      ...   
5105                       0                  1                        0   
5106                       0                  0                        1   
5107                       0                  0                        1   
5108                       0                  1                        0   
5109                       0                  0                        0   

      work_type_children  Residence_type_Rural  Residence_type_Urban  \
0                      0                     0                     1   
1                      0                     1                     0   
2                      0                     1                     0   
3                      0                     0                     1   
4                      0                     1                     0   
...                  ...                   ...                   ...   
5105                   0                     0                     1   
5106                   0                     0                     1   
5107                   0                     1                     0   
5108                   0                     1                     0   
5109                   0                     0                     1   

      smoking_status_Unknown  smoking_status_formerly smoked  \
0                          0                               1   
1                          0                               0   
2                          0                               0   
3                          0                               0   
4                          0                               0   
...                      ...                             ...   
5105                       0                               0   
5106                       0                               0   
5107                       0                               0   
5108                       0                               1   
5109                       1                               0   

      smoking_status_never smoked  smoking_status_smokes  
0                               0                      0  
1                               1                      0  
2                               1                      0  
3                               0                      1  
4                               1                      0  
...                           ...                    ...  
5105                            1                      0  
5106                            1                      0  
5107                            1                      0  
5108                        

In [ ]:
# Using Scikit Learn (Inferior)
gen_cat = df_cat_mdt_code['gender'].unique(); gen_cat
ohe = OneHotEncoder(handle_unknown='ignore')
ohe_fit = ohe.fit_transform(df_cat_mdt_code[['gender']])
# (Missing Data Treated) Dummy Dataset using Scikit Learn One Hot Encoder
df_mdt_gen_dummy_ohe = pd.DataFrame(ohe_fit.toarray(), columns='gender_'+gen_cat); df_mdt_gen_dummy_ohe

gender_Male  gender_Female  gender_Other
0             0.0            1.0           0.0
1             1.0            0.0           0.0
2             0.0            1.0           0.0
3             1.0            0.0           0.0
4             1.0            0.0           0.0
...           ...            ...           ...
5105          1.0            0.0           0.0
5106          1.0            0.0           0.0
5107          1.0            0.0           0.0
5108          0.0            1.0           0.0
5109          1.0            0.0           0.0

[5110 rows x 3 columns]

# **3. Data Transformation & Rescaling [Treatment of Outliers]**

**3.1. Standardization**

In [ ]:
ss = StandardScaler()
# Scaling Variable : bmi
ss_fit = ss.fit_transform(df_noncat_mdt[['bmi']])
df_noncat_std = pd.DataFrame(ss_fit, columns=['bmi_std']); df_noncat_std
#df_noncat_std = pd.DataFrame(ss_fit, columns=df_noncat_mdt.columns+'_std'); df_noncat_std
# (Missing Data Treated) Standardized Non-Categorical Dataset using Scikit Learn Standard Scaler
df_noncat_mdt_std = df_noncat_mdt.join(df_noncat_std); df_noncat_mdt_std
df_noncat_mdt_std = pd.merge(df_noncat_mdt, df_noncat_std, left_index=True, right_index=True); df_noncat_mdt_std


id   age  hypertension  heart_disease  ever_married  \
0      9046.0  67.0           0.0            1.0           1.0   
1     51676.0  61.0           0.0            0.0           1.0   
2     31112.0  80.0           0.0            1.0           1.0   
3     60182.0  49.0           0.0            0.0           1.0   
4      1665.0  79.0           1.0            0.0           1.0   
...       ...   ...           ...            ...           ...   
5105  18234.0  80.0           1.0            0.0           1.0   
5106  44873.0  81.0           0.0            0.0           1.0   
5107  19723.0  35.0           0.0            0.0           1.0   
5108  37544.0  51.0           0.0            0.0           1.0   
5109  44679.0  44.0           0.0            0.0           1.0   

      avg_glucose_level    bmi  stroke   bmi_std  
0                228.69  36.60     1.0  0.990618  
1                202.21  28.08     1.0 -0.110219  
2                105.92  32.50     1.0  0.460872  
3                171.23  34.40     1.0  0.706364  
4                174.12  24.00     1.0 -0.637381  
...                 ...    ...     ...       ...  
5105              83.75  32.40     0.0  0.447952  
5106             125.20  40.00     0.0  1.429919  
5107              82.99  30.60     0.0  0.215381  
5108             166.29  25.60     0.0 -0.430651  
5109              85.28  26.20     0.0 -0.353127  

[5110 rows x 9 columns]

**3.2. Normalization : Min-Max Scaling**

In [ ]:
mms = MinMaxScaler()
mms_fit = mms.fit_transform(df_noncat_mdt[['bmi']])
df_noncat_minmax_norm = pd.DataFrame(mms_fit, columns=['bmi_mmnorm']); df_noncat_minmax_norm
#df_noncat_minmax_norm = pd.DataFrame(mms_fit, columns=df_noncat_mdt.columns+'_mmnorm'); df_noncat_minmax_norm
# (Missing Data Treated) Normalized Non-Categorical Dataset using Sikit Learn Min-Max Scaler
df_noncat_mdt_mmn = df_noncat_mdt.join(df_noncat_minmax_norm); df_noncat_mdt_mmn
df_noncat_mdt_mmn = pd.merge(df_noncat_mdt, df_noncat_minmax_norm, left_index=True, right_index=True); df_noncat_mdt_mmn

id   age  hypertension  heart_disease  ever_married  \
0      9046.0  67.0           0.0            1.0           1.0   
1     51676.0  61.0           0.0            0.0           1.0   
2     31112.0  80.0           0.0            1.0           1.0   
3     60182.0  49.0           0.0            0.0           1.0   
4      1665.0  79.0           1.0            0.0           1.0   
...       ...   ...           ...            ...           ...   
5105  18234.0  80.0           1.0            0.0           1.0   
5106  44873.0  81.0           0.0            0.0           1.0   
5107  19723.0  35.0           0.0            0.0           1.0   
5108  37544.0  51.0           0.0            0.0           1.0   
5109  44679.0  44.0           0.0            0.0           1.0   

      avg_glucose_level    bmi  stroke  bmi_mmnorm  
0                228.69  36.60     1.0    0.301260  
1                202.21  28.08     1.0    0.203666  
2                105.92  32.50     1.0    0.254296  
3                171.23  34.40     1.0    0.276060  
4                174.12  24.00     1.0    0.156930  
...                 ...    ...     ...         ...  
5105              83.75  32.40     0.0    0.253150  
5106             125.20  40.00     0.0    0.340206  
5107              82.99  30.60     0.0    0.232532  
5108             166.29  25.60     0.0    0.175258  
5109              85.28  26.20     0.0    0.182131  

[5110 rows x 9 columns]

***3.3. Normalization : Robust Scaling ***

In [ ]:
# quantile_range=(25.0, 75.0) - Default Range
rs = RobustScaler(quantile_range=(10.0, 90.0))
rs_fit = rs.fit_transform(df_noncat_mdt[['bmi']])
df_noncat_robust_norm = pd.DataFrame(rs_fit, columns=['bmi_robnorm']); df_noncat_robust_norm
#df_noncat_robust_norm = pd.DataFrame(rs_fit, columns=df_noncat_mdt.columns+'_robnorm'); df_noncat_robust_norm
# (Missing Data Treated) Normalized Non-Categorical Dataset using Scikit Learn Robust Scaler
df_noncat_mdt_rn = df_noncat_mdt.join(df_noncat_robust_norm); df_noncat_mdt_rn
df_noncat_mdt_rn = pd.merge(df_noncat_mdt, df_noncat_robust_norm, left_index=True, right_index=True); df_noncat_mdt_rn

id   age  hypertension  heart_disease  ever_married  \
0      9046.0  67.0           0.0            1.0           1.0   
1     51676.0  61.0           0.0            0.0           1.0   
2     31112.0  80.0           0.0            1.0           1.0   
3     60182.0  49.0           0.0            0.0           1.0   
4      1665.0  79.0           1.0            0.0           1.0   
...       ...   ...           ...            ...           ...   
5105  18234.0  80.0           1.0            0.0           1.0   
5106  44873.0  81.0           0.0            0.0           1.0   
5107  19723.0  35.0           0.0            0.0           1.0   
5108  37544.0  51.0           0.0            0.0           1.0   
5109  44679.0  44.0           0.0            0.0           1.0   

      avg_glucose_level    bmi  stroke  bmi_robnorm  
0                228.69  36.60     1.0     0.444209  
1                202.21  28.08     1.0    -0.006346  
2                105.92  32.50     1.0     0.227393  
3                171.23  34.40     1.0     0.327869  
4                174.12  24.00     1.0    -0.222105  
...                 ...    ...     ...          ...  
5105              83.75  32.40     0.0     0.222105  
5106             125.20  40.00     0.0     0.624008  
5107              82.99  30.60     0.0     0.126917  
5108             166.29  25.60     0.0    -0.137493  
5109              85.28  26.20     0.0    -0.105764  

[5110 rows x 9 columns]

**3.4. Log Transformation**

In [ ]:
df_noncat_mdt_lt = df_noncat_mdt.copy()
# (Missing Data Treated) Log Tranformed Non-Categorical Dataset
df_noncat_mdt_lt['ln_bmi'] = np.log(df_noncat_mdt_lt['bmi']); df_noncat_mdt_lt
df_noncat_mdt_lt['log10_bmi'] = np.log10(df_noncat_mdt_lt['bmi']); df_noncat_mdt_lt

id   age  hypertension  heart_disease  ever_married  \
0      9046.0  67.0           0.0            1.0           1.0   
1     51676.0  61.0           0.0            0.0           1.0   
2     31112.0  80.0           0.0            1.0           1.0   
3     60182.0  49.0           0.0            0.0           1.0   
4      1665.0  79.0           1.0            0.0           1.0   
...       ...   ...           ...            ...           ...   
5105  18234.0  80.0           1.0            0.0           1.0   
5106  44873.0  81.0           0.0            0.0           1.0   
5107  19723.0  35.0           0.0            0.0           1.0   
5108  37544.0  51.0           0.0            0.0           1.0   
5109  44679.0  44.0           0.0            0.0           1.0   

      avg_glucose_level    bmi  stroke    ln_bmi  log10_bmi  
0                228.69  36.60     1.0  3.600048   1.563481  
1                202.21  28.08     1.0  3.335058   1.448397  
2                105.92  32.50     1.0  3.481240   1.511883  
3                171.23  34.40     1.0  3.538057   1.536558  
4                174.12  24.00     1.0  3.178054   1.380211  
...                 ...    ...     ...       ...        ...  
5105              83.75  32.40     0.0  3.478158   1.510545  
5106             125.20  40.00     0.0  3.688879   1.602060  
5107              82.99  30.60     0.0  3.421000   1.485721  
5108             166.29  25.60     0.0  3.242592   1.408240  
5109              85.28  26.20     0.0  3.265759   1.418301  

[5110 rows x 10 columns]

In [ ]:
# Pre-Processed Categorical Data Subset
df_cat_ppd = df_cat_mdt_code_oe.copy(); df_cat_ppd # Preferred Data Subset
# Pre-Processed Non-Categorical Data Subset
df_noncat_ppd = df_noncat_mdt_mmn.copy(); df_noncat_ppd # Preferred Data Subset
# Pre-Processed Dataset
df_ppd = df_cat_ppd.join(df_noncat_ppd); df_ppd # Pre-Processed Dataset
df_ppd = pd.merge(df_cat_ppd, df_noncat_ppd, left_index=True, right_index=True); df_ppd

gender      work_type Residence_type   smoking_status  gender_code  \
0       Male        Private          Urban  formerly smoked          1.0   
1     Female  Self-employed          Rural     never smoked          0.0   
2       Male        Private          Rural     never smoked          1.0   
3     Female        Private          Urban           smokes          0.0   
4     Female  Self-employed          Rural     never smoked          0.0   
...      ...            ...            ...              ...          ...   
5105  Female        Private          Urban     never smoked          0.0   
5106  Female  Self-employed          Urban     never smoked          0.0   
5107  Female  Self-employed          Rural     never smoked          0.0   
5108    Male        Private          Rural  formerly smoked          1.0   
5109  Female       Govt_job          Urban          Unknown          0.0   

      work_type_code  Residence_type_code  smoking_status_code       id   age  \
0                2.0                  1.0                  1.0   9046.0  67.0   
1                3.0                  0.0                  2.0  51676.0  61.0   
2                2.0                  0.0                  2.0  31112.0  80.0   
3                2.0                  1.0                  3.0  60182.0  49.0   
4                3.0                  0.0                  2.0   1665.0  79.0   
...              ...                  ...                  ...      ...   ...   
5105             2.0                  1.0                  2.0  18234.0  80.0   
5106             3.0                  1.0                  2.0  44873.0  81.0   
5107             3.0                  0.0                  2.0  19723.0  35.0   
5108             2.0                  0.0                  1.0  37544.0  51.0   
5109             0.0                  1.0                  0.0  44679.0  44.0   

      hypertension  heart_disease  ever_married  avg_glucose_level    bmi  \
0              0.0            1.0           1.0             228.69  36.60   
1              0.0            0.0           1.0             202.21  28.08   
2              0.0            1.0           1.0             105.92  32.50   
3              0.0            0.0           1.0             171.23  34.40   
4              1.0            0.0           1.0             174.12  24.00   
...            ...            ...           ...                ...    ...   
5105           1.0            0.0           1.0              83.75  32.40   
5106           0.0            0.0           1.0             125.20  40.00   
5107           0.0            0.0           1.0              82.99  30.60   
5108           0.0            0.0           1.0             166.29  25.60   
5109           0.0            0.0           1.0              85.28  26.20   

      stroke  bmi_mmnorm  
0        1.0    0.301260  
1        1.0    0.203666  
2        1.0    0.254296  
3        1.0    0.276060  
4        1.0    0.156930  
...      ...         ...  
5105     0.0    0.253150  
5106     0.0    0.340206  
5107     0.0    0.232532  
5108     0.0    0.175258  
5109     0.0    0.182131  

[5110 rows x 17 columns]

# **4. Data Bifurcation [Training & Testing Datasets]**

In [ ]:
train_df, test_df = train_test_split(df_ppd, test_size=0.25, random_state=1234)
# Training Dataset
train_df
# Testing Dataset
test_df, train_df

(      gender work_type Residence_type   smoking_status  gender_code  \
 3716  Female   Private          Urban           smokes          0.0   
 3129    Male  children          Urban          Unknown          1.0   
 3730    Male   Private          Urban           smokes          1.0   
 1610  Female   Private          Rural          Unknown          0.0   
 3512  Female   Private          Rural          Unknown          0.0   
 ...      ...       ...            ...              ...          ...   
 1947  Female   Private          Rural     never smoked          0.0   
 2125    Male   Private          Rural     never smoked          1.0   
 3741  Female   Private          Rural          Unknown          0.0   
 3102    Male   Private          Rural  formerly smoked          1.0   
 1592  Female   Private          Rural     never smoked          0.0   
 
       work_type_code  Residence_type_code  smoking_status_code       id   age  \
 3716             2.0                  1.0          

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
import os


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
train_df.to_csv('/content/drive/My Drive/Sample data/Treated_dataset.csv', index=False)


# **REPORT**

**DATA DESCRIPTION**

THE GIVEN DATA SET PROVIDES DETAILS ABOUT PEOPLE AND IF THEY HAVE SUFFERED FROM STROKE AND FACTORS THAT MIGHT OR MIGHT NOT AFFECT THE OCCURENCE. THE DATA SET CAN BE USED TO ANALYSE VARIOUS RELATIONSHIPS AND USE THESE RELATIONSHIPS TO PREDICT THE POSSIBILITY OF A STROKE USING VARIOUS MACHINE LEARNING ALGORITHMS.

**OBJECTIVES:**

1.TO IDENTIFY THE NUMBER OF NULL/EMPTY VALUES AND REMOVE THEM FROM THE DATASET.

2.TO CONVERT DATA TYPES TO PREPARE THE DATA FOR FURTHER ANALYSIS.

3.TO ADJUST NEW VALUES INTO EMPTY CELLS TO COMPLETE THE DATA.

4.TO TREAT THE OUTLIERS IN THE DATASET.

5.TO ENCODE THE DATASET TO MAKE IT EASY TO ANALYSE.

6.TO NORMALISE THE NON CATEGORICAL DATA  TO IMPROVE MACHINE LEARNING ALGORITHMS.

7.TO CONVERT THE DATA SET INTO NORMAL DISTRIBUTION TO GET BETTER ANALYSIS.

8.TO BIFURCATE THE DATA SET IN TRAINED AND TESTED.


**PROCEDURE:**

DATA PREPROCESSING STARTED WITH LOADING THE DATASET AND IDENTIFYING ITS COLUMNS. THEN, NULL VALUES FROM THE DATASET WAS IDENTIFIED AND WERE REPLACED BY THE CENTRAL TENDENCY MEASURES (CATEGORICAL- MEAN , NON CATEGORICAL- MEAN) TO MAKE THE DATA ELIGIBLE FOR FURTHER DATA ANALYSIS.THE DATA TYPES WERE CHANGED AND FURTHER TREATMENT WAS DONE TO REMOVE EMPTY VALUES. FINALLY,THE DATASET WAS ENCODED FOR BETTER ANALYSIS USING DECISION TREE AND OUTLIERS WERE TREATED TO PREPARE THE DATA FOR FINAL ANALYSIS.

**INSIGHTS:**

1.MISSING VALUES WAS THE MAXIMUM IN BMI IN THE DATASET I.E 201 VALUES.

2.


**MANAGERIAL IMPLICATIONS**

1.TO HELP MAKE DATA APPROPIATE FOR USING MACHINE LEARNING ALGORITHMS TO DRAW VALUABLE INSIGHTS.

2.TO PREDICT THE OCCURENCE OF STROKE IN PEOPLE BASED ON THEIR HABITS.

3.TO HELP MAKE ACCURATE AND PRECISE PREDICTION BASED ON DATA PATTERN.

4.TO IDENTIFY THE CAUSES AND THEIR LEVEL OF IMPACT ON OCCURENCE OF A STROKE.
